File: sentimental-credit/credit.py
from re import match


# TODO
def main():
    str_num = get_input()

    if not luhn_check(str_num):
        return print("INVALID\n")

    print_results(str_num)


def get_input():
    pattern = r"^\d{10,}$"

    while True:
        number = input("Number: ")

        if match(pattern, number):
            return number
        else:
            print("Invalid card number!")


def luhn_check(string):
    sum = 0
    reverse = string[::-1]

    for i in range(len(reverse)):
        if i % 2 != 0:
            num = int(reverse[i]) * 2

            for char in str(num):
                sum += int(char)

        else:
            sum += int(reverse[i])

    return sum % 10 == 0


def print_results(string):
    if len(string) == 15 and string[:2] in ["34", "37"]:
        print("AMEX")

    elif len(string) == 16 and 51 <= int(string[:2]) <= 55:
        print("MASTERCARD")

    elif len(string) in [13, 16] and string[:1] == "4":
        print("VISA")

    else:
        print("INVALID")


main()

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

int count_digits(long number);
long get_input(void);
int luhn_check(long number, int digits);
int get_first_digits(long number);
void print_result(int isValid, int firstDigits, int digits);

int main(void)
{
    long number = get_input();

    int digits = count_digits(number);

    int isValid = luhn_check(number, digits);

    int firstDigits = get_first_digits(number);

    print_result(isValid, firstDigits, digits);
}

long get_input(void)
{
    long number;

    do
    {
        number = get_long("Number: ");
    }
    while (number <= 1000000000);

    return number;
}

int count_digits(long number)
{
    int digits = 0;
    long n = number;

    while (n >= 1)
    {
        digits++;
        n /= 10;
    }

    return digits;
}

int luhn_check(long number, int digits)
{
    long n = number;
    int sum = 0;

    for (int i = 1; i <= digits; i++)
    {
        if (i % 2 == 0)
        {
            int x = n % 10 * 2;

            if (x > 9)
            {
                while (x > 0)
                {
                    sum += x % 10;
                    x /= 10;
                }
            }
            else
            {
                sum += x;
            }
        }
        else
        {
            sum += n % 10;
        }

        n /= 10;
    }

    return sum % 10 == 0;
}

int get_first_digits(long number)
{
    long n = number;

    while (n >= 100)
    {
        n /= 10;
    }

    return (int) n;
}

void print_result(int isValid, int firstDigits, int digits)
{
    if (isValid && digits == 15 && (firstDigits == 34 || firstDigits == 37))
    {
        printf("AMEX\n");
    }
    else if (isValid && digits == 16 && firstDigits >= 51 && firstDigits <= 55)
    {
        printf("MASTERCARD\n");
    }
    else if (isValid && (digits == 13 || digits == 16) && firstDigits >= 40 && firstDigits < 50)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}